controller/sync.go

```diff
@@ -275,6 +276,13 @@ func (sc *syncContext) sync() {
 	sc.log.WithFields(log.Fields{"tasks": tasks}).Debug("wet-run")
 	if !sc.runTasks(tasks, false) {
 		sc.setOperationPhase(v1alpha1.OperationFailed, "one or more objects failed to apply")
+	} else {
+		// if it is the last phase/wave and the only remaining tasks are non-hooks, the we are successful
+		// EVEN if those objects subsequently degraded
+		// this does create a special-case logic that we need to deal with
+		if len(preFilterTasks.Filter(func(t *syncTask) bool { return t.phase != phase || wave != t.wave() || t.isHook() })) == 0 {
```

Comment:
I think the logic should actually be:

`( t.isHook() && t.isRunnable() )`

Corner case: if I have a Deployment as an application hook (instead of a job/pod/workflow), and we are at the last wave, we should complete as soon as the apply was successful.

Reply:
I'm not sure about this. The `preFilterList` will only contain running tasks (line 613) or tasks that have not been run (I'm assuming this is what you mean by 'isRunnable'). I'm going to refactor this, as the code is complex.